### DISCUSSION TITLE
Mie Scattering tutorial: field instabilities with increasing resolution

### DISCUSSION THREAD

**Discussion by dalarev** [NanoComp/meep#1911] [Q&A] [answered]
Hi Everybody,

Can't figure out why the [Mie scattering example](https://github.com/NanoComp/meep/blob/master/python/examples/mie_scattering.py) becomes unstable at higher resolutions.

Running a copy/paste of the program produces the plot below, which is close enough (perhaps some outdated `dpml`, `dair`, or `sim.run(until_after_sources=#)` parameter is causing the mismatch at longer wavelengths?).

![mie_scattering_25](https://user-images.githubusercontent.com/52722332/150159025-f36915db-5106-4555-a70f-b5e53e811152.jpg)

Increasing the resolution from 25 to 30 produces the following. Things are already looking off. Y-limits are not hard-coded in the plot call, so I'm not sure why some MEEP results are off-screen.

![mie_scattering_30](https://user-images.githubusercontent.com/52722332/150159417-c3c78a6f-cb39-4ca1-8964-dd4fff5fa31e.jpg)

Increasing the resolution to 35 produces the following. MEEP results are completely off-screen, and again I'm not sure why, since there are no y-limits hard-coded.

![mie_scattering_35](https://user-images.githubusercontent.com/52722332/150159623-6c7d7d0b-8fdb-47f4-a34e-08ec981bed00.jpg)

Finally, `resolution=40` produces the following results. Results are again sensical.  

![mie_scattering_40](https://user-images.githubusercontent.com/52722332/150159738-2cbb408c-bcdc-4960-ad0d-702d7978f950.jpg)

Why could this be happening? It is a low index dielectric sphere. I ran into this issue trying to simulate Au nanospheres, where you need to crank up the res to account for the wavelengths inside the material.

I've tried extending the parameters mentioned above - no improvement. 

I have not tried turning off dielectric smoothing.

Any insight is appreciated!

**Comment by oskooi** [comment, marked as answer]
The cause of the field instability is due to the combination of the periodic boundary conditions (necessary for the planewave source excitation) and the two mirror symmetries as described in #132 (a long-standing bug). Removing the `symmetries` object in the script should fix the instabilities at the cost of increasing the problem size by a factor of four which can be an issue if you are performing a [convergence check](https://meep.readthedocs.io/en/latest/FAQ/#checking-convergence) and thus need to crank up the grid resolution, run time, and PML thickness. Also, the field instabilities are exacerbated by the presence of lossy materials (i.e., Au) which are, in general, more challenging to model in a Mie scattering calculation as I described in [#1530 (comment)](https://github.com/NanoComp/meep/issues/1530#issuecomment-1002192047).

(The parameters for `dpml`, `dair`, and the runtime of the scattering calculation in the [tutorial](https://meep.readthedocs.io/en/latest/Python_Tutorials/Basics/#mie-scattering-of-a-lossless-dielectric-sphere) are all *smaller* than what they need to be in order to reproduce the [plot](https://meep.readthedocs.io/en/latest/images/mie_scattering.png). I will update the tutorial with the correct numbers soon.)

**Reply by dalarev** [reply]
Awesome. Lots of useful information, thanks Oskooi.  

All suggestions (increase distance between geometry and flux box, increase `resolution`, remove `symmetries`) spell out much heavier computation costs - an inherent limitation of the algorithm, as you point out.

Is there a "rule of thumb" for how large the distance between the geometry and the flux planes should be? Presumably it's a function of your geometry size and wvl_max.

In general, perhaps outside of FDTD, what would you consider a better tool to simulate not just single NPs but NPs above a substrate, for example?

**Comment by dalarev** [comment]
Are periodic boundary conditions actually needed when using PMLs?

I removed `k_point` from the `sim` definition and reran the tutorial example (with `resolution=25`). Results are still slightly off at longer wavelengths, but there is strong correlation.

![mie_scattering_nokpoint](https://user-images.githubusercontent.com/52722332/150190475-e75f102d-bdf4-4f26-9069-e6c806c086c4.jpg)

Setting `resolution=30` previously produced very different results (2nd image in OP). Now, with `k_point` removed, results are still looking good, though it's tough to say if this is an improvement over `resolution=25`.

![mie_scattering](https://user-images.githubusercontent.com/52722332/150191576-84c29986-765f-4777-ac99-6c6e4833d8c6.jpg)

These results are all with symmetries still enabled.

**Reply by oskooi** [reply]
In this particular example, because the planewave source extends into the PML all the way to the edge of the cell it does not matter what the boundary conditions are (assuming the PML is thick enough). This is why, as you have discovered, removing the periodic boundary conditions (and replacing them with the default metallic boundaries) does not affect the results. The main cause of the discrepancy at large wavelengths between the Meep results and the Mie theory is likely the runtime which needs to be sufficiently long to ensure that the DFT fields are converged. Apart from the runtime, you should also verify that the results are converged with respect to the PML thickness (`dpml`) and the distance between the line source and the (nearest) DFT monitor (`dair`). Note that you typically need to repeatedly *double* the simulation parameters to verify convergence.
